aten/src/ATen/DeviceGuard.h

```diff
@@ -6,42 +6,55 @@
 #include <ATen/detail/CUDAHooksInterface.h>
 #include <c10/util/Exception.h>
 #include "c10/util/Optional.h"
+#include <c10/detail/DeviceGuardImplInterface.h>
 
 #include <cstddef>
 
 namespace at {
 /// RAII guard that sets a certain default device in its constructor, and
-/// changes it back to the device that was originally active upon destruction.
+/// changes it back to the device (for that device type) that was originally
+/// active upon destruction.
 ///
-/// The device is always reset to the one that was active at the time of
-/// construction of the guard. Even if you `set_device` after construction, the
-/// destructor will still reset the device to the one that was active at
-/// construction time.
-struct DeviceGuard {
-  /// Default constructor, does nothing.
-  DeviceGuard() = default;
-
+/// If the device is changed via this guard to a different one than the
+/// active one at construction time, this guard will reset it to the one
+/// that was active at the time of construction of the guard.  WARNING: if
+/// you change the current device out-of-band, e.g., by directly calling
+/// cudaSetDevice(), DeviceGuard is NOT guaranteed to reset it upon
+/// exiting this scope.  The contract required by DeviceGuard is that inner code
+/// leaves the device in the same state that DeviceGuard set it.  In DEBUG mode,
+/// we check for this invariant.
+///
+/// If a DeviceGuard is constructed without specifying a device type (this
+/// can occur if you, e.g., pass a nullopt to the constructor), it behaves as if
+/// it were a no-op "CPU" guard; e.g., current_device() reports that the current
+/// device is kCPU.  This is different from passing Device(kCUDA, -1), which
+/// says to use the current CUDA device; in this case, we will correctly query
+/// what the current CUDA device is, but won't change it.
+class DeviceGuard {
+public:
   /// Set the current device to the passed Device.
   explicit DeviceGuard(Device device) {
-    set_device(device);
+    init_device(device);
   }
 
-  explicit DeviceGuard(c10::optional<Device> device_opt) {
+  /// Set the current device to the passed Device, if not nullopt;
+  /// otherwise do nothing.
+  explicit DeviceGuard(optional<Device> device_opt) {
```

Comment:
Do we need this special case? Can't, instead, the call site unwrap the optional and call device guard with the value? Seems to just complicate things, even needing a full paragraph of explanation in the class comment.

Reply:
It seems to be pretty convenient. Here's the problem: you want to stack allocate. Let's say you don't have this constructor. You write this:

```
if (device_opt.has_value()) {
  DeviceGuard guard(device_opt.value());
  // Wait uh....
}
the_actual_compute();
```

The DeviceGuard needs to be promoted to the top level somehow. You used to be able to do this with `set_device`, but on DeviceGuard `set_device` is hugely problematic for other reasons.

I suppose you could write it like this using the move constructor.

```
DeviceGuard guard;
if (device_opt.has_value()) {
  guard = DeviceGuard(device_opt.value());
}
```

...actually, I think this is totally broken right now, in the current implementation.

Reply:
what about

    auto guard = device_opt.map( [] (auto opt) {return DeviceGuard(opt); });

Reply:
Or, since c++17 std::optional doesn't have `::map()` yet (meh), it's probably either going to be

    auto guard = map(device_opt, [] (auto opt) {return DeviceGuard(opt); });

or with the move constructor as you wrote above, that's likely fine too.

Reply:
optional<Device> comes up naturally -- for example it's how we parse Devices from the python arg parser, when passing in None is valid.  It's way nicer to operate directly on that than to write the expressions above.

Reply:
map() is a bit too magical, I think the constructor with optional is more readable if it's a common use case